src/liger_kernel/ops/cross_entropy.py

```diff
@@ -73,16 +75,36 @@ def liger_cross_entropy_kernel(
         d = d * tl.exp(m - m_new) + tl.sum(tl.exp(X_block - m_new))
         m = m_new
 
+    # we need to compute sum(log(softmax(x_i))) for smooth_loss
+    smooth_loss = 0.0
+    log_d = tl.log(d)  # avoid redundant calculations
+    if label_smoothing > 0:
+        for i in range(0, n_cols, BLOCK_SIZE):
+            X_offsets = i + tl.arange(0, BLOCK_SIZE)
+            X_block = tl.load(
+                X_ptr + X_offsets,
+                mask=X_offsets < n_cols,
+                other=(
+                    m + log_d
+                ),  # out-of-bounds will become 0 after calculating softmax
+            )
+            smooth_loss += -tl.sum(X_block - m - log_d)
+
     # 4. [Online softmax] second pass: calculate the gradients
     # dx_y = (softmax(x_y) - 1) / N
     # dx_i = softmax(x_i) / N, i != y
     # N is the number of non ignored elements in the batch
+    # For label smoothing:
+    # dx_i = (softmax(x_y) - label_smoothing / V) / N, V = n_cols, i != y
+    # dx_y = (softmax(x_y) - label_smoothing / V - (1 - label_smoothing)) / N
+    #      = dx_i - (1 - label_smoothing) / N
+    eps = label_smoothing / n_cols
```

Comment:
concerned about naming

Reply:
Comments for label smoothing version are identical to non label smoothing ones.
Not sure which one we should keep, keeping both for now.